elastic/apm-server
# Introduce dedicated fields for internal metrics

## Motivation/summary

Add dedicated fields to model.Transaction and model.Span for well-defined, internal transaction and span metrics. Drop unknown metrics received in metricsets with any transaction or span fields defined.

Internal metrics will be indexed with strict mapping. We already have explicitly defined fields for these metrics. This PR ensures that unknown metrics do not slip through. If an agent is upgraded before the server and some new internal metric starts being produced, that must not be indexed as it would cause other internal metrics to be rejected by Elasticsearch.

Drop the unused `metricset.period` field that was being added to service_destination metrics.

## Checklist

~- [ ] Documentation has been updated~

## How to test these changes

- Run apm-server with `-E instrumentation.enabled=true`
- Check that breakdown metrics (Time spent by span type) work correctly in the UI
- Check that service destination metrics (Dependencies) work correctly in the UI

## Related issues

- https://github.com/elastic/apm-server/issues/5084
- https://github.com/elastic/apm-server/issues/4410
- https://github.com/elastic/apm-server/pull/6043#issuecomment-905257392
